### DISCUSSION TITLE
Issue having CMS as a module, can the Orchard CMS security being reused on Modules?

### DISCUSSION THREAD

**Discussion by gabrielacosta** [OrchardCMS/OrchardCore#15842] [Q&A]
Hi, 

I'm trying to use Orchard and have:

1. Landing project using CMS
2. Module Admin - using Module
3. Module customer - using Module. 

I create a solution with a working example of what I'm trying to achieve. 
https://github.com/gabrielacosta/WebApplication1

First, I created everything and it works fine. I added to the landing only the OrchardCore.Module.Targets Nuget. 

Landing

![image](https://user-images.githubusercontent.com/8519425/112561090-9edc3600-8e28-11eb-9fe7-5af3be46d6d7.png)

Admin

![image](https://user-images.githubusercontent.com/8519425/112561119-ac91bb80-8e28-11eb-8d70-7faa8267f090.png)

Customer

![image](https://user-images.githubusercontent.com/8519425/112561142-b74c5080-8e28-11eb-9809-08d932861b13.png)
Until that point, everything works fine. 

However, when I tried to add the CMS Nuget to the Landing Module. 

![image](https://user-images.githubusercontent.com/8519425/112561279-04302700-8e29-11eb-9a73-28d0197a926f.png)

And modify the startup on the landing to load OrchardCMS

![image](https://user-images.githubusercontent.com/8519425/112561409-42c5e180-8e29-11eb-8531-e7a4be1b6d37.png)

The solution compiles without any issues. 

However, when I run it I got this error. 

![image](https://user-images.githubusercontent.com/8519425/112561481-6c7f0880-8e29-11eb-9c35-6f17cf1b999e.png)

I'm not sure how to fix it. Or if the model that I'm trying to build is correct. 

Could you please review the solution and let me know. 

Kind regards, 

G

**Comment by Skrypt** [comment]
This is a bad idea to reference Orchard Core inside a module. It should be referenced only on your main website project like OrchardCore.Cms.Web for example in our templates.

If you need to reference a specific OrchardCore module in your project you can reference these instead.

If I look at your solution folders and project I'd say that OrchardCore.Cms.Web would be equivalent to your Main.csproj. This is the only place where you should reference OrchardCore.Application.Cms.Target.

**Comment by Skrypt** [comment]
Please explain what you are trying to achieve so that we can understand better and help you. Else the solution is to revert the changes that makes this fail to run.

**Comment by gabrielacosta** [comment]
Hi Skrypt, 
Thanks for replying. 
What I'm trying to achieve is having one website as landing using the Orchard CMS. Then I want to add to it more modules that will do other functionalities, but still use security model from the CMS. 
So, if have two specific roles such as admin and customer, those modules can have dashboards or other CRUD type of code not related with the content of the landing website.

**Comment by Skrypt** [comment]
Create 2 different **tenants** then.

Landing Website = host tenant
Child Website = activate module 1
Child Website 2 = activate module 2 ...

Then these 2 websites/tenants are not going to use the same Content Types than the host one. And your modules can register a custom route for your "/dashboard" which will be secured by it's controller like we do in OrchardCore.Contents module AdminController. But instead of naming it AdminController you name it DashboardController.

And this way each tenants can have their own custom dashboard. But you should not activate both modules in the same tenant if they both register a /dashboard route.

**Comment by gabrielacosta** [comment]
Hi Skrypt, 

I just tried. 

![image](https://user-images.githubusercontent.com/8519425/112569085-a5be7500-8e37-11eb-9b7b-b8dcec23ef6a.png)

And I got another error. 

![image](https://user-images.githubusercontent.com/8519425/112569131-b8d14500-8e37-11eb-809d-74b64a788e25.png)

I checked other possible issues, but I'm quite sure I do not use AddMvc() in the startup of the landing. 

![image](https://user-images.githubusercontent.com/8519425/112569210-dc948b00-8e37-11eb-9ff8-14807b6e1dda.png)

Do you know, what could be the issue. 

Kind regards,
G

**Comment by Skrypt** [comment]
Look at the templates for examples : 

https://github.com/OrchardCMS/OrchardCore/tree/dev/src/Templates/OrchardCore.ProjectTemplates/content

Your CMSLanding here will be probably needing the startup.cs and program.cs from this template module : 

https://github.com/OrchardCMS/OrchardCore/tree/dev/src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Cms.Web

**Comment by gabrielacosta** [comment]
I already the startup and the program as recommended from the templates. 

https://github.com/OrchardCMS/OrchardCore/blob/dev/src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Cms.Web/Program.cs

and 
https://github.com/OrchardCMS/OrchardCore/blob/dev/src/Templates/OrchardCore.ProjectTemplates/content/OrchardCore.Templates.Cms.Web/Startup.cs

However, the error still appearing...

![image](https://user-images.githubusercontent.com/8519425/112570106-87f20f80-8e39-11eb-8175-a22987ace4c0.png)

I think another ticket was setup before https://github.com/OrchardCMS/OrchardCore/issues/3453

But, I do not use Sqlite, the landing CMS is using the blank template and SQL Server 2012.

I upload the solution to 

https://github.com/gabrielacosta/CMSwithModules

**Comment by deanmarcussen** [comment]
Your module / startup definitions don't seem correct.

Suggest:

Create one Web project which calls `.UseOrchardCore()`

Create as many modules as you want after that. 
Do your code in those modules. 
Don't call `.UseOrchardCore()` again.

**Comment by gabrielacosta** [comment]
Hi Dean,

Thanks for the suggestion. I already try that and I know it works. However, I want to use the cms component. I don't get how your suggestion helps me to achieve that.

**Comment by weirdyang** [comment]
> Hi Dean,
> 
> Thanks for the suggestion. I already try that and I know it works. However, I want to use the cms component. I don't get how your suggestion helps me to achieve that.

@gabrielacosta, 
The issue is you have not created the tenant, but your appsettings.json has configuration for it with the states set to `running`. but if you inspect the `tenants.json` in your App_Data folder, there is only the `default` tenant. Note; this is only an issue if you're using OrchardCMS

create a new occms 
first install the templates: 
`dotnet new -i OrchardCore.ProjectTemplates::1.0.0-*`
then navigate to a folder:
`dotnet new occms -n MySite`
launch the site, select `blog` or `saas`
navigate to features and activate `tenants` feature:

![image](https://user-images.githubusercontent.com/25576022/116880856-189be680-ac55-11eb-9780-bb7589c40d60.png)

do the setup:
![image](https://user-images.githubusercontent.com/25576022/116880946-3d905980-ac55-11eb-9e82-2c2652681d3a.png)

click setup:
![image](https://user-images.githubusercontent.com/25576022/116880992-4bde7580-ac55-11eb-93d1-5b1c6bbdee96.png)

navigate to the prefix you set:
![image](https://user-images.githubusercontent.com/25576022/116881074-64e72680-ac55-11eb-9df3-72c55aa33488.png)

![image](https://user-images.githubusercontent.com/25576022/116881231-94962e80-ac55-11eb-8830-e060e16f24d0.png)

your App_Data folder should now have a new entry in `tenants.json`.

**Comment by weirdyang** [comment]
See https://github.com/OrchardCMS/OrchardCore/pull/5422 for how to preconfigure it:

> So now, for the documentation
> 
> To provide preconfig values for a given tenant you need a tenant valid state as Uninitialized. This will automatically create a tenant at runtime using these values.
> 
> To only provide config values for a given tenant don't specify a tenant state. These values will be only used when the tenant is manually created.

here's a screenshot; 

![image](https://user-images.githubusercontent.com/25576022/116882315-e68b8400-ac56-11eb-8e65-1edbde089db8.png)

Hope that helps!

Here's the repo; https://github.com/weirdyang/orchardcore-demo
